server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java

```diff
@@ -185,15 +185,39 @@ private void downloadToFS(String fileName, Path location, String primaryTerm) th
     }
 
     public TranslogTransferMetadata readMetadata() throws IOException {
-        return transferService.listAll(remoteMetadataTransferPath).stream().max(METADATA_FILENAME_COMPARATOR).map(filename -> {
-            try (InputStream inputStream = transferService.downloadBlob(remoteMetadataTransferPath, filename)) {
-                IndexInput indexInput = new ByteArrayIndexInput("metadata file", inputStream.readAllBytes());
-                return metadataStreamWrapper.readStream(indexInput);
-            } catch (IOException e) {
-                logger.error(() -> new ParameterizedMessage("Exception while reading metadata file: {}", filename), e);
-                return null;
-            }
-        }).orElse(null);
+        SetOnce<TranslogTransferMetadata> metadataSetOnce = new SetOnce<>();
+        SetOnce<IOException> exceptionSetOnce = new SetOnce<>();
+        final CountDownLatch latch = new CountDownLatch(1);
+        LatchedActionListener<List<BlobMetadata>> latchedActionListener = new LatchedActionListener<>(
+            ActionListener.wrap(blobMetadataList -> {
+                if (blobMetadataList.isEmpty()) return;
+                String filename = blobMetadataList.get(0).name();
+                try (InputStream inputStream = transferService.downloadBlob(remoteMetadataTransferPath, filename)) {
+                    IndexInput indexInput = new ByteArrayIndexInput("metadata file", inputStream.readAllBytes());
+                    metadataSetOnce.set(metadataStreamWrapper.readStream(indexInput));
+                } catch (IOException e) {
+                    logger.error(() -> new ParameterizedMessage("Exception while reading metadata file: {}", filename), e);
+                    exceptionSetOnce.set(e);
+                }
+            }, e -> {
+                logger.error(() -> new ParameterizedMessage("Exception while listing metadata files "), e);
+                exceptionSetOnce.set((IOException) e);
+            }),
+            latch
+        );
+
+        try {
+            transferService.listAllInSortedOrder(remoteMetadataTransferPath, 1, latchedActionListener);
+            latch.await();
```

Comment:
Do we need to keep a max timeout?

Reply:
I would rather rely on Blob Store to timeout using idle connection or do retries internally . Because if we timeout , shard is going to fail and is going to get retried .